js-packages/@prestojs/util/src/useAsync.ts

```diff
@@ -0,0 +1,439 @@
+ *          return r.json();
+ *      }
+ *      throw r;
+ *   });
+ * }
+ * function FollowerCount() {
+ *     const [user, setUser] = React.useState('octocat')
+ *     const { response, isLoading, error, run, reset } = useAsync(() => getGithubUser(user));
+ *     return (
+ *         <div>
+ *             <InputWidget value={user} onChange={e => setUser(e.target.value)} />
+ *             <Button onClick={run} loading={isLoading}>Query follower count</Button>
+ *             <Button onClick={reset}>Clear</Button>
+ *             {response && (
+ *                 <p>
+ *                     <img src={response.avatar_url} /><br />
+ *                     {response.name} has {response.followers} followers
+ *                 </p>
+ *             )}
+ *             {error && (<p>Failed with status: {error.status} {error.statusText}</p>)}
+ *         </div>
+ *     );
+ * }
+ * <FollowerCount />
+ * ```
+ * @param fn A function that returns a promise. When `trigger` is `MANUAL` this is only
+ * called when you manually call the returned `run` function, otherwise it's called
+ * initially and then whenever an equality comparison fails between previous arguments and new
+ * arguments.
+ *
+ *
+ * @extract-docs
+ */
+function useAsync<ResponseT, ErrorT>(
+    fn: (...args: Array<any>) => Promise<ResponseT>,
+    options: UseAsyncOptions = {}
+): UseAsyncReturnObject {
+    const { trigger = MANUAL, args = [], onSuccess, onError } = options;
+    const [state, dispatch] = useReducer<
+        (
+            state: ReducerState<ResponseT, ErrorT>,
+            action: ReducerAction<ResponseT, ErrorT>
+        ) => ReducerState<ResponseT, ErrorT>
+    >(reducer, {
+        isLoading: trigger !== MANUAL,
+        error: null,
+        response: null,
+    });
+
+    // =========================================================================
+    // Validate parameters
+    const invalidOptionKeys = Object.keys(options).filter(key => !validOptionKeys.includes(key));
+    if (invalidOptionKeys.length > 0) {
+        throw new Error(
+            `Invalid options specified: ${invalidOptionKeys.join(
+                ', '
+            )}. Valid options are: ${validOptionKeys.join(', ')}`
+        );
+    }
+    if (options && !isPlainObject(options)) {
+        throw new Error(`Options must be an object, received ${options}`);
+    }
+    if (args && !Array.isArray(args)) {
+        throw new Error(
+            `The 'args' option must be an array of arguments to pass to the async function. If you are trying to pass a single argument wrap it in an array, eg. '[<arg>]'. Received type: ${typeof args}`
+        );
+    }
+    if (![SHALLOW, MANUAL, DEEP].includes(trigger)) {
+        throw new Error(`Invalid trigger ${trigger}`);
+    }
+    // =========================================================================
+
+    const comparisonFn = comparisonByTrigger[trigger];
+    const memoizedArgs = useMemoOne(() => args, args, comparisonFn);
```

Comment:
Is useMemoOne really necessary? If some results are re-fetched then does that matter? (Should this behaviour be an optional param?)

Reply:
it's necessary for the `comparisonFn` support - without it it would change every time

Reply:
`comparisonFn` is a reference into a global table `comparisonByTrigger` -- it be stable between renders

Reply:
edit: or do you mean that it's the existences of `comparisonFn` that is needed, and that the `One` part of `useMemoOne` is actually irrelevant?

Reply:
yes but `useMemo` doesn't support a custom comparison - we need `useMemoOne` (or something like it) for that